### ISSUE TITLE
Regression Model is not converging on my custom data

### ISSUE THREAD

**Issue by ShikaiWang-Sky** [google-deepmind/regress-lm#57]
Hi, thanks a lot for your great work on RegressLM!

I’m currently experimenting with my own dataset. The same dataset can be regressed well using a simple MLP, but when I try it with RegressLM, the training never converges.

Here is an example input (key-value pair format):
```
{
  "instruction": "Predict the DC gain of the circuit given the device sizing parameters.",
  "input": {
    "M1_L": 7.1, "M1_W": 1.6, "M1_M": 41.0,
    "M3_L": 2.1, "M3_W": 7.1, "M3_M": 66.0,
    "M5_L": 2.1, "M5_W": 2.6, "M5_M": 31.0,
    "M6_L": 5.1, "M6_W": 3.1, "M6_M": 11.0,
    "M7_L": 6.6, "M7_W": 3.6, "M7_M": 76.0,
    "M8_L": 6.6, "M8_W": 2.6, "M8_M": 81.0,
    "C1_L": 51.0, "C1_W": 81.0,
    "Ib": 2.1e-05
  }
}
```

Target value:

89.87658

My observations:

I tried fine-tuning with ProtectedSentencePieceVocab including all parameter keys.

I also tried without protection. Both cases do not converge.

Training loss stays above ~8 and doesn’t decrease.

When I evaluate using the saved checkpoint, predictions collapse to a constant close to the global mean.

Do you have any suggestions on what I might be missing (e.g., tokenization setup, hyperparameters, data formatting)?

Any help or guidance would be greatly appreciated!

Thanks again for sharing this great project.

**Comment by ShikaiWang-Sky** [comment]
> Hey, some pointers:
> 
> 1. I'm assuming that whole dict (with both instruction and input) will be turned into a string right? Since `x` needs to be a string finally.
> 2. You don't need the full sentence "Predict the DC gain..." for the "instruction" since the random-init model doesn't understand English anyways, just some bare-bones identifier. If you're not doing multi-task regression, then this instruction isn't needed at all.
> 3. Make sure your optimizer and learning rate are tuned properly. You should at least see on the training data, that the cross-entropy loss goes down. At least, you'd need to make sure you fit the training data (disregarding the test).
> 4. If the predictions collapse to the global mean, this means the model is having a hard time reading the input. Double check the input tokenization / representation being sent in.

Hi, thanks again for your helpful reply!

I double-checked the input dict x and confirmed that it is indeed converted into a string before feeding into the model. To eliminate any potential bias from the tokenizer, I also simplified the input format. For example, my current inputs look like this:
```
{
    "input": {
        "A": "2.4",
        "B": "3.8",
        "C": "35",
        "D": "4.8",
        "E": "4.9",
        "F": "61",
        "G": "6",
        "H": "3.3",
        "I": "80",
        "J": "1.4",
        "K": "6.1",
        "L": "2",
        "M": "1.7",
        "N": "5.6",
        "O": "70",
        "P": "1.7",
        "Q": "1.5",
        "R": "76",
        "S": "99",
        "T": "74",
        "U": "4e-05"
    }
}
```
However, the results are still similar: even when I scale up the dataset, the validation loss decreases (e.g., from ~19 down to ~8) but then stabilizes at around 8 and does not go further. In contrast, a simple MLP baseline on the same dataset converges much better.

This makes me wonder: could you possibly share a small example dataset (or even just a snippet) that has been successfully used to fine-tune the regression LM? Having a reproducible dataset where regression works end-to-end would be extremely helpful as a reference, to make sure my current data format and training setup are aligned with the intended usage.

For completeness, here is my current fine-tuning script (kept close to the default code, with minimal modifications):
```
from regress_lm import core
from regress_lm import rlm
from torch import optim
from regress_lm import rlm
from regress_lm import vocabs
from regress_lm import tokenizers
from regress_lm.models.pytorch import model as pytorch_model
import pandas as pd
from pathlib import Path
from regress_lm.core import Example
import torch

# LOAD DATA
# load train csv
train_df = pd.read_csv("labels/train.csv")
test_df = pd.read_csv("labels/test.csv")

# load netlist text
format_type = "sizing"
if format_type == "spice":
    train_dir = Path("rendered_netlists/train")
    test_dir = Path("rendered_netlists/test")
elif format_type == "kv":
    train_dir = Path("kv_netlist/train")
    test_dir = Path("kv_netlist/test")
elif format_type == "sizing":
    train_dir = Path("pure_sizing/train")
    test_dir = Path("pure_sizing/test")
else:
    raise ValueError(f"Invalid format type: {format_type}")

x_texts_train = []
y_targets_train = []
x_texts_test = []
y_targets_test = []

for idx, row in train_df.iterrows():
    if format_type == "spice":
        netlist_path = train_dir / f"opamp1_{idx}.spice"
    elif format_type == "kv":
        netlist_path = train_dir / f"opamp1_{idx}.kv.txt"
    elif format_type == "sizing":
        netlist_path = train_dir / f"opamp1_{idx}.json"
    if netlist_path.exists():
        x_texts_train.append(netlist_path.read_text())
        # TODO: use multiple targets
        y_targets_train.append(float(row["dcgain"]))

for idx, row in test_df.iterrows():
    if format_type == "spice":
        netlist_path = test_dir / f"opamp1_{idx}.spice"
    elif format_type == "kv":
        netlist_path = test_dir / f"opamp1_{idx}.kv.txt"
    elif format_type == "sizing":
        netlist_path = test_dir / f"opamp1_{idx}.json"
    if netlist_path.exists():
        x_texts_test.append(netlist_path.read_text())
        y_targets_test.append(float(row["dcgain"]))

# construct training examples
examples_train = [Example(x=x, y=y) for x, y in zip(x_texts_train, y_targets_train)]
examples_test = [Example(x=x, y=y) for x, y in zip(x_texts_test, y_targets_test)]

# Create RegressLM with max input token length.
params_dict = {
    "max_input_len": 2048,
    "batch_size": 256,
}
reg_lm = rlm.RegressLM.from_default(device="cuda:3", **params_dict)

# Example (x,y) pairs, which can be fine-tuned against.
reg_lm.fine_tune(examples_train, validation_examples=examples_test, seed=42)
torch.save(reg_lm.model.state_dict(), "checkpoints/best_only_weights.pt")
```
I’d greatly appreciate any guidance on what the “right” data format should look like for successful fine-tuning. Thanks a lot for your time and support!

**Comment by xingyousong** [comment]
Hi in this case - what i suspect is that training the RLM from scratch isn't enough for your specific data. 

Training a good checkpoint over massive amounts of synthetic tabular data first is one of our TODOs, but no guarantees on when it can be done - but this is what I'd recommend next, if you're able to. It would allow the model to better understand the continuity / formatting of your problem.

By this I mean e.g. training over {"A": 1.0, "B": 3.5", ...} where the inputs correspond to inputs to a cheap synthetic function (like a sphere function).

**Comment by ShikaiWang-Sky** [comment]
> Hi in this case - what i suspect is that training the RLM from scratch isn't enough for your specific data.
> 
> Training a good checkpoint over massive amounts of synthetic tabular data first is one of our TODOs, but no guarantees on when it can be done - but this is what I'd recommend next, if you're able to. It would allow the model to better understand the continuity / formatting of your problem.
> 
> By this I mean e.g. training over {"A": 1.0, "B": 3.5", ...} where the inputs correspond to inputs to a cheap synthetic function (like a sphere function).

Hi, thanks again for your reply!

I tried to fine-tune the model on a trivial function y = x, but the loss gets stuck around ~11, which is really strange. This task should be extremely easy for the model to learn.

For your convenience, here is a minimal reproducible example:
```
# rlm_x2_mre_noleak_kv_default.py
import numpy as np
import torch
from pathlib import Path

from regress_lm.core import Example
from regress_lm import rlm

torch.set_float32_matmul_precision('high')
device = "cuda:0" if torch.cuda.is_available() else "cpu"
rng = np.random.default_rng(42)

# ---------- 1) Disjoint integer splits (NO data leakage) ----------
X_MIN, X_MAX = -3000, 3000
N_TRAIN, N_VAL, N_TEST = 2560, 640, 2560
assert (X_MAX - X_MIN + 1) >= (N_TRAIN + N_VAL + N_TEST), "Domain too small for disjoint splits."

all_ints = np.arange(X_MIN, X_MAX + 1, dtype=int)
rng.shuffle(all_ints)

x_train = all_ints[:N_TRAIN]
x_val   = all_ints[N_TRAIN:N_TRAIN + N_VAL]
x_test  = all_ints[N_TRAIN + N_VAL:N_TRAIN + N_VAL + N_TEST]

# Ground-truth function (float for safety)
y_train = (x_train.astype(int))
y_val   = (x_val.astype(int))
y_test  = (x_test.astype(int))

# ---------- 2) KV input format: "X1: <int>" ----------
def fmt_kv(x_int: int) -> str:
    return f"X1: {x_int}"

examples_train = [Example(x=fmt_kv(x), y=int(t)) for x, t in zip(x_train, y_train)]
examples_val   = [Example(x=fmt_kv(x), y=int(t)) for x, t in zip(x_val,   y_val)]
examples_test  = [Example(x=fmt_kv(x), y=int(t)) for x, t in zip(x_test,  y_test)]

# ---------- 3) Build RegressLM with DEFAULT optimizer/fine-tuner ----------
reg = rlm.RegressLM.from_default(
    device=device,
    max_input_len=64,   # small, since inputs are short KV strings
    compile_model=False # disable compile for reproducibility/debug
)

print("first 10 examples in train set:")
for e in examples_train[:10]:
    print(type(e.x), type(e.y))
    print(e.x, e.y)
    
print("first 10 examples in val set:")
for e in examples_val[:10]:
    print(e.x, e.y)
    
print("first 10 examples in test set:")
for e in examples_test[:10]:
    print(e.x, e.y)

# ---------- 4) Train on train; validate on val (defaults) ----------
print("=== Train on disjoint integer train set; validate on disjoint val set ===")
reg.fine_tune(examples_train, validation_examples=examples_val, seed=123)

# ---------- 5) Inference on test ----------
def predict_float(regressor: rlm.RegressLM, examples):
    ys = regressor.sample([Example(x=e.x, y=0.0) for e in examples], num_samples=1)
    preds = np.array([float(y.squeeze()) for y in ys], dtype=float)
    return preds

pred_test = predict_float(reg, examples_test)

# ---------- 6) Evaluation ----------
mae = np.mean(np.abs(pred_test - y_test))
rmse = np.sqrt(np.mean((pred_test - y_test)**2))
ss_res = np.sum((pred_test - y_test)**2)
ss_tot = np.sum((y_test - np.mean(y_test))**2)
r2 = 1 - ss_res/ss_tot

print(f"[TEST] MAE={mae:.6f}  RMSE={rmse:.6f}  R2={r2:.6f}")

# ---------- 7) Save weights ----------
Path("checkpoints").mkdir(exist_ok=True, parents=True)
torch.save(reg.model.state_dict(), "checkpoints/rlm_x2_kv_default.pt")
print("Saved to checkpoints/rlm_x2_kv_default.pt")
```
Training behavior:
Training loss only decreases from ~20 to ~10–11 and then stalls.
Validation/test metrics remain very poor.
For such a trivial mapping, I expected near-zero error (R² ≈ 1).

Could you help confirm whether this is a bug or if I’m missing something in how targets should be represented?
